lib/internal/http2/core.js

```diff
@@ -717,6 +719,17 @@ function setupHandle(socket, type, options) {
 
   this[kHandle] = handle;
 
+  if (socket.encrypted) {
+    this[kAlpnProtocol] = socket.alpnProtocol;
+    this[kEncrypted] = true;
+    const origin =
+      (new URL(`https://${socket.servername}:${socket.remotePort}`)).origin;
```

Comment:
Can we avoid parsing a URL for every connected socket?

Reply:
To my knowlege sockets can be also connected via Unix Domain sockets/windows pipes. There is no remotePort in that case.

Reply:
Good point. In those cases, there won't be a proper initial `origin`. I'll update to account for that.

Reply:
@mcollina ... the ORIGIN spec is pretty clear that the `origin` must be a validly serialized ASCII-origin, which means our `servername` must be a pct-encoded ASCII host. The spec also requires that the SNI servername is used to construct the `origin`. I can see if we can make this more efficient, but I don't think we can completely avoid this.

We *would* potentially defer the parse such that it occurs the first time the `originSet` property getter is accessed.

Reply:
> We would potentially defer the parse such that it occurs the first time the originSet property getter is accessed.

I like that idea. 👍

Reply:
>  the ORIGIN spec is pretty clear that the origin must be a validly serialized ASCII-origin, which means our servername must be a pct-encoded ASCII host. The spec also requires that the SNI servername is used to construct the origin. I can see if we can make this more efficient, but I don't think we can completely avoid this.

I'm guessing we do not need to generate a new string and then parse it to generate a new pct-encoded ASCII host.

> We could potentially defer the parse such that it occurs the first time the originSet property getter is accessed.

+1 with this approach.

Reply:
Unfortunately, as far as I know, the `servername` isn't guaranteed to be ASCII or percent encoded so we need to at least pass that through.

Reply:
> Unfortunately, as far as I know, the servername isn't guaranteed to be ASCII or percent encoded so we need to at least pass that through.

can that be cached and filled in from the server?